types/project.go

```diff
@@ -353,3 +357,49 @@ func (p *Project) ResolveImages(resolver func(named reference.Named) (godigest.D
 	}
 	return eg.Wait()
 }
+
+// ResolveServicesEnvironment parse env_files set for services to resolve the actual environment map for services
+func (p Project) ResolveServicesEnvironment(discardEnvFiles bool) error {
+	for i, service := range p.Services {
+		service.Environment = service.Environment.Resolve(p.Environment.Resolve)
+
+		environment := MappingWithEquals{}
+		// resolve variables based on other files we already parsed, + project's environment
+		var resolve dotenv.LookupFn = func(s string) (string, bool) {
+			v, ok := environment[s]
+			if ok && v != nil {
+				return *v, ok
+			}
+			return p.Environment.Resolve(s)
+		}
+
+		for _, envFile := range service.EnvFile {
+			file, err := os.Open(envFile)
+			if err != nil {
+				return errors.Wrapf(err, "Failed to read %s", file)
+			}
+
+			b, err := io.ReadAll(file)
+			if err != nil {
+				return errors.Wrapf(err, "Failed to read %s", file)
+			}
+
+			// Do not defer to avoid it inside a loop
+			file.Close() //nolint:errcheck
```

Comment:
This will leak if we end up returning an error above - perhaps `os.ReadFile` instead?

Reply:
+1, or if we just want to pass file name we should use `envFile` variable directly